enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,613 @@
+
+**Failed restore on rollback (cont of "simple host reboot")**
+
+1. Second ostree commit is staged
+1. First ostree commit shuts down
+1. Second ostree commit boots
+1. MicroShift pre-run
+   - Backup mode is "backup",
+   - Backup script runs, creating a backup compatible with the first ostree commit
+1. MicroShift data migration or startup fail (upgrade was allowed)
+1. Greenboot runs red scripts
+    - Set backup mode to "restore"
+1. Second ostree commit shuts down
+
+1. Following repeats `GREENBOOT_MAX_BOOT_ATTEMPTS - 1` times
+   1. Second ostree commit boots
+   1. MicroShift pre-run
+      - Backup mode is "restore",
+      - Backup compatible with the first ostree commit is restored
+   1. MicroShift data migration or startup fail (upgrade was allowed)
+   1. Greenboot runs red scripts
+      - Set backup mode to "restore"
+
+1. Greenboot issues rollback (too many failed boots)
+1. Second ostree commit shuts down
+1. First ostree commit boots
+1. MicroShift pre-run
+   - Backup mode is "restore"
+   - Backup tool **fails to restore the backup** compatible with the first ostree commit
+1. MicroShift cluster doesn't start
+1. MicroShift / healthcheck detects that restore after rollback
+1. MicroShift greenboot healthcheck is overridden to return success
+   - Endless reboot and rollback loop is avoided
+1. Manual intervention is needed
+
+Otherwise following will happen:
+
+12. MicroShift's greenboot healthcheck fails
+0. Greenboot runs red scripts
+    - Set backup mode to "restore" and reboot the host
+0. (Failures repeat and rollback is issued)
+0. **First (!!!)** ostree commit shuts down
+0. **Second (!!!)** ostree commit boots
+0. Repeats ad infinitum
+
+**Fail first startup, FDO (FIDO Device Onboard) deployment**
+
+1. An ostree commit without MicroShift is installed on the device at the factory.
+1. The device boots at a customer site.
+1. An agent in the ostree commit performs FIDO device onboarding or a similar process to determine the workload.
+1. An ostree commit with MicroShift installed is staged.
+1. The sans-MicroShift commit shuts down.
+1. The with-MicroShift commit starts up.
+1. MicroShift starts
+   - No backup action (did not previously run, so there is nothing to backup, and did not previously fail so no restore needed)
+1. MicroShift startup fails
+1. Greenboot runs red scripts
+   - Set backup mode to "restore"
+1. (Failures may need to repeat to trigger rollback.)
+1. The with-MicroShift ostree commit shuts down
+1. The sans-MicroShift ostree commit boots
+1. The agent stages with-MicroShift ostree commit 2
+1. The sans-MicroShift ostree commit shuts down
+1. The with-MicroShift ostree commit 2 starts up.
+1. MicroShift starts
+   - The backup mode is "restore"
+   - There is no backup.
+   - (Open Question) XOR
+     - Remove MicroShift data and start with clean slate
+     - Try running with existing data
+       - What if greenboot fails? Deleting data allows fresh start
```

Comment:
Yes, it would give us more data to work on:

- Green boot, green MicroShift
- Red boot, green MicroShift -> no need to restore (won't lose any data), but not sure if we should persist `backup` or `restore`
- Red boot, red MicroShift (thing to remember: we might not be exclusively at fault here) -> restore